### PR TITLE
Create backup action

### PR DESCRIPTION
**Pull request by javierdelapuente**
Applicable spec: ISD095

### Overview

This PR creates the action `create-backup` to back up Synapse, using the s3 integration to get
S3 credentials and the `backup_passphrase` config variable to use as the encryption
key. 

To create a backup, a tar file is created with all the files that are required for the
backup, then it is encrypted and sent to a S3 compatible object storage. All these
operations are done in the Synapse container using container.exec with a shell
command using pipes, like `tar -c <paths> | gpg --symmetric <options> | aws s3 cp <options>`.  For
that command to work, the gpg `passphrase` to use was previously copied to the 
container filesystem as a file and the aws client configured with the necessary options and 
environment variables.

Besides, the aws s3 command has an option to give the expected size of the object to upload (`--expected-size` ). This is required so the aws s3 cp can calculate the correct size for the multipart uploads. This is done in this PR using a bare `du` command with all the paths to back up.

### Rationale

Synapse data and config directories should be backed up for disaster and recovery purposes.

### Juju Events Changes

Added `create-backup` action.

### Module Changes

The entry point for the logic is the `backup_observer`, which register a handler for the `create-backup` action.

All the rest of the logic is inside the `backup` module, that performs several actions in the workload to 
stream the backup, as explained in the overview section.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/backup.py [hunk 1]
```diff
@@ -3,18 +3,47 @@
...
+
+# For the data directory, inside the "media" directory, all directories starting
+# with local_ will be backed up. The directories starting with "remote_" are from
+# other server and is it not necessary to back them up.
+MEDIA_DIR = "media"
+LOCAL_DIR_PATTERN = "local_*"
+
+# A smaller value will minimise memory requirements. A bigger value can make the transfer faster.
+S3_MAX_CONCURRENT_REQUESTS = 1
+PASSPHRASE_FILE = "/root/.gpg_passphrase"  # nosec
```
Comment (javierdelapuente) [inline src/backup.py:35]: Not sure this is the right place. Any idea?

Comment (arturo-seijas) [reply src/backup.py:35]: Do you mean the key location? This would imply running synapse as root user, which contravenes the least privilege principle

Comment (arturo-seijas) [reply src/backup.py:35]: Perhaps under the application directory (`/srv/synapse/...` I guess) or under `$HOME` (wherever that is)?

Comment (javierdelapuente) [reply src/backup.py]: I have updated the location of the passphrase file to Synapse config dir. I haven't changed to run it as the Synapse user as it is managed by the startup script and it is not fully configured (has no home dir for example).

WDYT?

**Review by arturo-seijas — DISMISSED** [review]

File: src/backup.py [hunk 2]
```diff
@@ -128,3 +157,203 @@ def can_use_bucket(self) -> bool:
...
+        paths: Paths to check.
+
+    Returns:
+        Total size in bytes.
+
+    Raises:
+        BackupError: If there was a problem calculating the size.
+    """
+    command = "set -euxo pipefail; du -bsc " + paths_to_args(paths) + " | tail -n1 | cut -f 1"
+    exec_process = container.exec([BASH_COMMAND, "-c", command])
```
Comment (javierdelapuente) [inline src/backup.py:207]: try catch

**Comment by javierdelapuente** [comment]
closed as it will be done in another branch

**Pull request closed without merging**